File: bindings/zig/src/destination.zig
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

const std = @import("std");
const c = @import("c.zig");
const types = @import("types.zig");
const util = @import("util.zig");

pub fn create(
    node: types.Node,
    identity: types.Identity,
    app_name: []const u8,
    aspects: []const []const u8,
    accepts_links: bool,
) types.Error!types.Destination {
    if (app_name.len == 0) return error.InvalidArg;

    var app_buf: [256]u8 = undefined;
    const app_z = try util.requireCString(app_name, &app_buf);

    var aspect_storage: [8][128]u8 = undefined;
    var aspect_ptrs: [8][*:0]const u8 = undefined;
    if (aspects.len > aspect_ptrs.len) return error.InvalidArg;
    for (aspects, 0..) |a, i| {
        aspect_ptrs[i] = try util.requireCString(a, aspect_storage[i][0..]);
    }

    const h = c.rns_destination_create(
        types.asU64(node),
        types.asU64(identity),
        app_z,
        if (aspects.len == 0) null else aspect_ptrs[0..aspects.len].ptr,
        aspects.len,
        if (accepts_links) 1 else 0,
    );
    if (h == 0) return error.Internal;
    return @enumFromInt(h);
}

pub fn enableRatchets(destination: types.Destination, path: ?[]const u8) types.Error!void {
    if (path) |p| {
        var path_buf: [4096]u8 = undefined;
        const path_z = try util.requireCString(p, &path_buf);
        try types.mapCode(c.rns_destination_enable_ratchets(types.asU64(destination), path_z));
        return;
    }
    try types.mapCode(c.rns_destination_enable_ratchets(types.asU64(destination), null));
}

pub fn enforceRatchets(destination: types.Destination) types.Error!void {
    try types.mapCode(c.rns_destination_enforce_ratchets(types.asU64(destination)));
}

pub fn announce(destination: types.Destination, app_data: []const u8) types.Error!void {
    if (app_data.len == 0) {
        try types.mapCode(c.rns_destination_announce(types.asU64(destination), null, 0));
        return;
    }
    try types.mapCode(c.rns_destination_announce(types.asU64(destination), app_data.ptr, app_data.len));
}

pub fn hash(destination: types.Destination) types.Error![types.hash_len]u8 {
    var out: [types.hash_len]u8 = undefined;
    var written: usize = 0;
    try types.mapCode(c.rns_destination_hash(types.asU64(destination), &out, out.len, &written));
    if (written != types.hash_len) return error.Truncated;
    return out;
}

pub fn destroy(destination: types.Destination) types.Error!void {
    try types.mapCode(c.rns_destination_destroy(types.asU64(destination)));
}

pub fn registerRequestHandler(destination: types.Destination, path: []const u8) types.Error!void {
    var path_buf: [256]u8 = undefined;
    const path_z = try util.requireCString(path, &path_buf);
    try types.mapCode(c.rns_destination_register_request_handler(types.asU64(destination), path_z));
}

File: bindings/rust/src/destination.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

use std::ffi::CString;
use std::ptr;

use crate::error::{map_code, Error, Result};
use crate::ffi::{self, HASH_LEN};
use crate::identity::Identity;
use crate::node::Node;

pub struct Destination {
    handle: u64,
}

impl Destination {
    pub fn create(
        node: &Node,
        identity: Option<&Identity>,
        app_name: &str,
        aspects: &[&str],
        accepts_links: bool,
    ) -> Result<Self> {
        if app_name.is_empty() {
            return Err(Error::InvalidArg);
        }
        let app = CString::new(app_name).map_err(|_| Error::InvalidArg)?;
        let mut aspect_ptrs: Vec<*const i8> = Vec::with_capacity(aspects.len());
        let mut aspect_strings: Vec<CString> = Vec::with_capacity(aspects.len());
        for aspect in aspects {
            aspect_strings.push(CString::new(*aspect).map_err(|_| Error::InvalidArg)?);
            aspect_ptrs.push(aspect_strings.last().unwrap().as_ptr());
        }
        let id_handle = identity.map(|i| i.handle()).unwrap_or(0);
        let h = unsafe {
            ffi::rns_destination_create(
                node.handle(),
                id_handle,
                app.as_ptr(),
                if aspect_ptrs.is_empty() {
                    ptr::null()
                } else {
                    aspect_ptrs.as_ptr()
                },
                aspect_ptrs.len(),
                if accepts_links { 1 } else { 0 },
            )
        };
        if h == 0 {
            return Err(Error::Internal);
        }
        Ok(Self { handle: h })
    }

    pub fn enable_ratchets(&self, path: Option<&str>) -> Result<()> {
        match path {
            None => map_code(unsafe { ffi::rns_destination_enable_ratchets(self.handle, ptr::null()) }),
            Some(p) => {
                let c = CString::new(p).map_err(|_| Error::InvalidArg)?;
                map_code(unsafe { ffi::rns_destination_enable_ratchets(self.handle, c.as_ptr()) })
            }
        }
    }

    pub fn enforce_ratchets(&self) -> Result<()> {
        map_code(unsafe { ffi::rns_destination_enforce_ratchets(self.handle) })
    }

    pub fn announce(&self, app_data: &[u8]) -> Result<()> {
        map_code(unsafe {
            ffi::rns_destination_announce(
                self.handle,
                if app_data.is_empty() {
                    ptr::null()
                } else {
                    app_data.as_ptr()
                },
                app_data.len(),
            )
        })
    }

    pub fn hash(&self) -> Result<[u8; HASH_LEN]> {
        let mut out = [0u8; HASH_LEN];
        let mut written = 0usize;
        map_code(unsafe {
            ffi::rns_destination_hash(self.handle, out.as_mut_ptr(), out.len(), &mut written)
        })?;
        if written != HASH_LEN {
            return Err(Error::Truncated);
        }
        Ok(out)
    }

    pub fn register_request_handler(&self, path: &str) -> Result<()> {
        if path.is_empty() {
            return Err(Error::InvalidArg);
        }
        let path = CString::new(path).map_err(|_| Error::InvalidArg)?;
        map_code(unsafe { ffi::rns_destination_register_request_handler(self.handle, path.as_ptr()) })
    }

    pub fn handle(&self) -> u64 {
        self.handle
    }
}

impl Drop for Destination {
    fn drop(&mut self) {
        if self.handle != 0 {
            unsafe {
                let _ = ffi::rns_destination_destroy(self.handle);
            }
            self.handle = 0;
        }
    }
}
